Wolfssl code back

jpbland1 opened the pull request:
## Summary
Add wolfSSL build files, allows wolfSSL to be cloned in and used quickly

## Impact
Allows wolfSSL to be built into NuttX

## Testing
Tested using wolfSSL tests in the NuttX simulator and on the STM32F746ZG, STM32H753 and STM32L552ZE

xiaoxiang781216 commented on crypto/wolfssl/user_settings.h:
```diff
@@ -0,0 +1,65 @@
+#define WOLF_C99
+
+/* Math */
+#if defined(CONFIG_ARCH_CHIP_STM32L552ZE) || defined(CONFIG_ARCH_CHIP_STM32F746ZG) || defined(CONFIG_ARCH_CHIP_STM32H743ZI)
```
why check this in common code

xiaoxiang781216 commented on crypto/wolfssl/Makefile:
```diff
@@ -0,0 +1,152 @@
+
+endif
+
+$(WOLFSSL_ZIP):
```
let's skip the download if .git exist in wolfssl

xiaoxiang781216 commented on crypto/wolfssl/Makefile:
```diff
@@ -0,0 +1,152 @@
+# wolfSSL SSL/TLS crypto library
+
+# Set up build configuration and environment
+https://github.com/wolfSSL/wolfssl/archive/refs/tags/v5.5.4-stable.zip
```
comment out?

xiaoxiang781216 commented on crypto/wolfssl/README.md:
```diff
@@ -76,110 +76,6 @@ Skip to step 6
-      |  ^~~~~~~
-LD: nuttx
-CP: nuttx.hex
-CP: nuttx.bin
```
let's merge the 2nd and 3rd patch to first one.

jpbland1 commented:
I've squashed the readme commits, removed the old zip and curl logic since we now clone directly from git and removed the platform specific checks in user_settings.h

xiaoxiang781216 commented on crypto/wolfssl/Makefile:
```diff
@@ -0,0 +1,136 @@
+CSRCS += $(WOLFSSL_UNPACKNAME)/wolfcrypt/src/wolfevent.c
+CSRCS += $(WOLFSSL_UNPACKNAME)/wolfcrypt/src/wolfmath.c
+
+
```
remove the blank line

xiaoxiang781216 commented on crypto/wolfssl/Makefile:
```diff
@@ -0,0 +1,136 @@
+
+CFLAGS += -DWOLFSSL_USER_SETTINGS
+
+
```
remove too

jpbland1 replied:
that has to be there, wolfssl needs that flag so it knows to read the user_settings.h file instead of the normal configure command

jpbland1 replied:
or did you mean the white space, that I removed

jpbland1 commented:
yeah sorry, should have just made it simple, squashed again

xiaoxiang781216 commented:
@jpbland1 but I don't see the download/clone in Makefile.

acassis commented:
ping @jpbland1

acassis commented on crypto/wolfssl/Kconfig:
```diff
@@ -0,0 +1,108 @@
+# see the file kconfig-language.txt in the NuttX tools repository.
+#
+
+menuconfig CRYPTO_WOLFSSL
+	bool "wolfSSL SSL/TLS Cryptography Library"
+	default n
+	---help---
+		Enable support for wolfSSL
```
I think wolfSSL should depends on CONFIG_ALLOW_GPL_COMPONENTS because is it not Apache license

xiaoxiang781216 replied:
@jpbland1 how about this comment?

jpbland1 replied:
@xiaoxiang781216 is that not the right way to do it?

xiaoxiang781216 replied:
Yes, but GPL is an invasive license, we need a way to highlight it to the user.

jpbland1 replied:
What do I need to do then? It already won't show up unless you enable GPL in the license select, would you like me to add something to the option text to make it obvious?

hartmannathan replied:
> Yes, but GPL is an invasive license, we need a way to highlight it to the user. Anyway, I am fine without ALLOW_GPL_COMPONENTS. What's your opinion @acassis and @hartmannathan ?

Just make it depends on ALLOW_GPL_COMPONENTS. Users have to enable that explicitly because it is False by default. If they enable it, then they accept the GPL license will apply to their binaries.

jpbland1 replied:
@hartmannathan I'm confused and maybe don't know enough about the build system, as it is currently the Kconfig option is wrapped with `if ALLOW_GPL_COMPONENTS` do I need to add something else to exclude it?

hartmannathan replied:
> @hartmannathan I'm confused and maybe don't know enough about the build system, as it is currently the Kconfig option is wrapped with `if ALLOW_GPL_COMPONENTS` do I need to add something else to exclude it?

@jpbland1 Apologies for my confusing reply earlier. I was replying to @xiaoxiang781216, but on second reading, it looks like I misunderstood the comment.

I think this PR is fine the way it is because it does show as "Depends on: ALLOW_GPL_COMPONENTS" in kconfig-frontends.

I tested as follows:

```
$ cd ../apps/
$ wget -q https://github.com/apache/nuttx-apps/pull/1557.patch
$ git apply 1557.patch
$ cd ../nuttx/
$ tools/configure.sh tm4c129e-launchpad:nsh
$ make menuconfig
```

If I do not select ALLOW_GPL_COMPONENTS then CRYPTO_WOLFSSL is invisible in the menu:

```
    [ ] LibTomCrypt CrypographyLibrary  ----
    [ ] Mbed TLS Cryptography Library  ----
        Crypto: TinyCrypt cryptography library  --->
```

I can search for CRYPTO_WOLFSSL and it shows the GPL dependency: 

```
Symbol: CRYPTO_WOLFSSL [=n]
Type  : boolean
Prompt: wolfSSL SSL/TLS Cryptography Library
  Location:
    -> Application Configuration
(1)   -> Cryptography Library Support
  Defined at /home/x/NuttX/apps/crypto/wolfssl/Kconfig:7
  Depends on: ALLOW_GPL_COMPONENTS [=n]
```

If I select ALLOW_GPL_COMPONENTS then CRYPTO_WOLFSSL appears in the menu:

```
    [ ] LibTomCrypt CrypographyLibrary  ----
    [ ] Mbed TLS Cryptography Library  ----
        Crypto: TinyCrypt cryptography library  --->
    [ ] wolfSSL SSL/TLS Cryptography Library (NEW)  ----
```

The help text for this item shows the dependency.

I think that's fine.

If we want to have "depends on ALLOW_GPL_COMPONENTS" actually written in the stanza of menuconfig CRYPTO_WOLFSSL, it can be done with a simple refactoring: just remove the outermost "if ALLOW_GPL_COMPONENTS" ... "endif # ALLOW_GPL_COMPONENTS" and add "depends on ALLOW_GPL_COMPONENTS" under menuconfig CRYPTO_WOLFSSL. It will have the same effect as now. The only difference is that perhaps some people prefer that as more self-documenting? More future-proof against this Kconfig file being edited in the future? I don't have a strong opinion either way. I'm perfectly OK with either way.

xiaoxiang781216 replied:
@jpbland1 sorry, I don't see you change at the begin, the patch looks fine now, let's merge it, thank you.

pkarashchenko commented on crypto/wolfssl/README.md:
````diff
@@ -0,0 +1,98 @@
+Clone wolfssl into the the the crypto/wolfssl folder
+
+```
+git clone https://github.com/wolfSSL/wolfssl.git apps/crypto/wolfssl/wolfssl
````
But what about proprietary license? I mean if someone wants to integrate NuttX with a paid library? We need to think about how to deal with dual-licensed code. Maybe add also ALLOW_PROPRIATERY_COMPONENTS or have a separate option to select it per component?

jpbland1 commented:
Finally got back to this, I didn't realize that getting the source code was supposed to be an automated step in the Makefile, I've also added the GPL allowed condition to the Kconfig, let me know if something else is wrong.

xiaoxiang781216 commented:
@jpbland1 please fix the below nxstyle warning:
```
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/crypto/wolfssl/user_settings.h:1:1: error: Missing file header comment block
Warning: /home/runner/work/nuttx-apps/nuttx-apps/apps/crypto/wolfssl/user_settings.h:1:1: warning: #include outside of 'Included Files' section
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/crypto/wolfssl/user_settings.h:31:1: error: Missing blank line after comment
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/crypto/wolfssl/user_settings.h:32:0: error: Missing space after opening C comment
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/crypto/wolfssl/user_settings.h:32:24: error: Missing space before closing C comment
```
and rebase to the last mainline.

acassis commented:
> @jpbland1 please fix the below nxstyle warning:
> 
> ```
> Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/crypto/wolfssl/user_settings.h:1:1: error: Missing file header comment block
> Warning: /home/runner/work/nuttx-apps/nuttx-apps/apps/crypto/wolfssl/user_settings.h:1:1: warning: #include outside of 'Included Files' section
> Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/crypto/wolfssl/user_settings.h:31:1: error: Missing blank line after comment
> Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/crypto/wolfssl/user_settings.h:32:0: error: Missing space after opening C comment
> Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/crypto/wolfssl/user_settings.h:32:24: error: Missing space before closing C comment
> ```
> 
> and rebase to the last mainline.

@jpbland1 just add the missing sections separation in the code and squash it and I will merge your PR!
